import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/backupuploader/GcpBackupUploader.java

```diff
@@ -1,25 +1,129 @@
 
 public class GcpBackupUploader extends AbstractBackupUploader {
+  private String projectName;
+  private String bucketName;
+  private String uploadObject;
+  private Storage storage;
 
-  public GcpBackupUploader(String uploadFilePath) {
+  private static final String ENV_CREDS = "GOOGLE_APPLICATION_CREDENTIALS";
+
+  public GcpBackupUploader(String uploadFilePath) throws IOException {
     super(uploadFilePath);
+    parseUploadFilePath(getUploadFilePath());
+    storage = getStorage();
+  }
+
+  public GcpBackupUploader(String uploadFilePath, Storage storage) throws IOException {
+    super(uploadFilePath);
+    parseUploadFilePath(getUploadFilePath());
+    this.storage = storage;
   }
 
   @Override
-  public void doWriteBackup(InputStream inputStream, String uniqueFileName) throws BackupException {
-    
+  public void doWriteBackup(InputStream backupData, String uniqueFileName)
+          throws BackupException {
+    try {
+      if (backupData == null) {
+        throw new BackupException("Backup data is null");
+      }
+      validatePathParameter(uniqueFileName, "unique file name");
+      BlobId blobId = BlobId.of(bucketName, getFullUploadObject(uniqueFileName));
+      BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+      storage.createFrom(blobInfo, backupData);
+    } catch (Exception e) {
+      throw new BackupException("Error with writing backup file: " + e.getMessage(), e);
+    }
   }
 
   @Override
   public InputStream doReadBackup(String uniqueFileName) throws BackupException {
-    return null;
+    try {
+      validatePathParameter(uniqueFileName, "unique file name");
+      Blob blob = storage.get(BlobId.of(bucketName, getFullUploadObject(uniqueFileName)));
+      ReadChannel channel = blob.reader();
+      return Channels.newInputStream(channel);
+    } catch (Exception e) {
+      throw new BackupException("Error with reading backup file: " + e.getMessage(), e);
+    }
   }
 
   @Override
   public void doRemoveBackup(String uniqueFileName) throws BackupException {
+    try {
+      validatePathParameter(uniqueFileName, "unique file name");
+      storage.delete(bucketName, getFullUploadObject(uniqueFileName));
+    } catch (Exception e) {
+      throw new BackupException("Error with removing backup file: " + e.getMessage(), e);
+    }
+  }
+
+  private void parseUploadFilePath(String uploadFilePath) throws GcpUriParseException {
+    try {
+      validatePathParameter(uploadFilePath, "upload file path");
+      List<String> segments = new URIBuilder().setPath(getUploadFilePath()).getPathSegments();
+      projectName = segments.get(2);
+      bucketName = segments.get(3);
+      uploadObject = new URIBuilder()
+              .setPathSegments(segments.subList(4, segments.size()))
+              .getPath().substring(1);
+      validatePathParameter(projectName, "project name");
+      validatePathParameter(bucketName, "bucket name");
+      validatePathParameter(uploadObject, "upload object");
```

Comment:
Isn't it more correct to call this the folder rather than the upload object?

Also, traditionally GCS file paths are "gs://bucketID/path/to/object", the project ID isn't included in the GCS URI. Could we split the GCS backup project ID into a separate flag (something like --gcs_backup_project_id ) so we can re-use the established concept of the GCS URI without any changes here?  The readme would also need to be updated.

Reply:
The documentation notes that the GCS structure includes a bucket and objects, the concept of a folder is not suitable for GCS -> https://cloud.google.com/storage/docs/introduction

Reply:
Ok what about 'object path prefix' or something like that?

Reply:
it looks like this is the name, not the path to it, because the name of the object may look like /level1/level2/some_file -> https://cloud.google.com/storage/docs/json_api/v1/objects